rir/src/compiler/translations/pir_2_rir.cpp

```diff
@@ -720,133 +645,141 @@ size_t Pir2Rir::compileCode(Context& ctx, Code* code) {
-            next instruction, and it is its only input, and PirCopy is not
-            involved
-            TODO: is it always the case that the store and load use the same
-            local slot?
-        */
-        auto store = [&](BB::Instrs::iterator it, Value* what) {
-            if (alloc.dead(what)) {
-                cs << BC::pop();
-                return;
-            }
-            if (alloc.onStack(what))
-                return;
-            cs << BC::stloc(alloc[what]);
-        };
-        auto load = [&](BB::Instrs::iterator it, Value* what) {
-            if (alloc.onStack(what))
-                return;
-            cs << BC::ldloc(alloc[what]);
-        };
-
         Value* currentEnv = nullptr;
-        auto loadEnv = [&](BB::Instrs::iterator it, Value* val) {
-            assert(Env::isAnyEnv(val));
-            if (Env::isStaticEnv(val))
-                cs << BC::push(Env::Cast(val)->rho);
-            else
-                load(it, val);
-        };
-        auto setEnv = [&](BB::Instrs::iterator it, Value* val) {
-            assert(Env::isAnyEnv(val));
-            if (val != currentEnv) {
-                currentEnv = val;
-                loadEnv(it, val);
-                cs << BC::setEnv();
-            } else if (alloc.allocation[val] == SSAAllocator::stackSlot) {
-                cs << BC::pop();
-            }
-        };
 
         for (auto it = bb->begin(); it != bb->end(); ++it) {
             auto instr = *it;
+
+            bool hasResult =
+                instr->type != PirType::voyd() && !Phi::Cast(instr);
+
+            auto explicitEnvValue = [](Instruction* instr) {
+                return MkEnv::Cast(instr) || Deopt::Cast(instr);
+            };
+
+            // Load Arguments to the stack
+            {
+                auto loadEnv = [&](BB::Instrs::iterator it, Value* what) {
+                    if (Env::isStaticEnv(what)) {
+                        cs << BC::push(Env::Cast(what)->rho);
+                    } else if (what == Env::notClosed()) {
+                        cs << BC::callerEnv();
+                    } else {
+                        if (!alloc.hasSlot(what)) {
+                            std::cerr << "Don't know how to load the env ";
+                            what->printRef(std::cerr);
+                            std::cerr << " (" << tagToStr(what->tag) << ")\n";
+                            assert(false);
+                        }
+                        if (!alloc.onStack(what))
+                            cs << BC::ldloc(alloc[what]);
+                    }
+                };
+
+                auto loadArg = [&](BB::Instrs::iterator it, Instruction* instr,
+                                   Value* what) {
+                    if (what == Missing::instance()) {
+                        // if missing flows into instructions with more than one
+                        // arg we will need stack shuffling here
```

Comment:
Do you have an example?

Reply:
well, let's assume you have `foo(a,missing,c)`, now you have a and c on the stack and you need to push missing in the middle.